### PR TITLE
improve error when name and parameters are not space-separated

### LINKED ISSUE #8934: Omitting a space in custom command declaration causes misleading error
### Describe the bug
Omitting a space between the function identifier and `[` causes a `Missing required positional argument` error
```nu
# test.nu

def test_1[] {
  echo "generates error"
}

def test_2 [] {
  echo "runs fine"
}
```

### How to reproduce

type `def test[] { echo "" }` into the REPL

### Expected behavior

Either the error should make sense given the context, or the tokenizer should accept this syntax

### Screenshots

image

### Configuration

| key                | value                                                            |
| ------------------ | ---------------------------------------------------------------- |
| version            | 0.78.0                                                           |
| branch             | makepkg                                                          |
| commit_hash        | 2ec20286371f1d90a77fdd5d82818c6c5851f090                         |
| build_os           | linux-x86_64                                                     |
| build_target       | x86_64-unknown-linux-gnu                                         |
| rust_version       | rustc 1.68.2 (9eb3afe9e 2023-03-27) (Arch Linux rust 1:1.68.2-1) |
| cargo_version      | cargo 1.68.2                                                     |
| build_time         | 2023-04-05 02:09:35 +00:00                                       |
| build_rust_channel | release                                                          |
| features           | default, zip                                                     |
| installed_plugins  |                                                                  |

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by 1Kinoti**
# Description
closes #8934

this pr improves the diagnostic emitted when the name and parameters of either `def`, `def-env` or `extern` are not separated by a space

```nu
Error:
  × no space between name and parameters
   ╭─[entry #1:1:1]
 1 │ def err[] {}
   ·        ▲
   ·        ╰── expected space
   ╰────
  help: consider adding a space between the `def` command's name and its parameters
```

from

```nu
Error: nu::parser::missing_positional

  × Missing required positional argument.
   ╭─[entry #1:1:1]
 1 │ def err[] {}
   ╰────
  help: Usage: def <def_name> <params> <body>
```

### CODE REVIEW

**Comment by jntrnr** [comment]
Should be detect this and then just let them not have to put the space if they don't want to? 🤔

**Comment by jntrnr** [comment]
@1Kinoti I wonder if we could check how many args were passed in, and if it's one less than what we expect, we look to see if we can convert the `def foo[] {}` to `def foo [] {}` (and same for `def foo() {}`)

**Comment by cmidkiff87** [comment]
> i haven't figure how to do that yet,sweat_smile since `def`s are parsed twice? paired with completions, i think it would difficult to do that. but it's possible,

Does Nushell not use a tokenizer before parsing? I feel like a tokenizer would make this trivial (but i have zero experience with a REPL or tokenizer, so...)

**Comment by 1Kinoti** [comment]
> I wonder if we could check how many args were passed in, and if it's one less than what we expect, we look to see if we can convert the def foo[] {} to def foo [] {} (and same for def foo() {})

@jntrnr where would we check this? in `parse_predecl()` or in `parse_def()`? or in both? i think this is possible

**Comment by 1Kinoti** [comment]
> Does Nushell not use a tokenizer before parsing? I feel like a tokenizer would make this trivial (but i have zero experience with making a REPL or tokenizer, so...)

@cmidkiff87 there **is** a [tokenizer](https://github.com/nushell/nushell/blob/main/crates/nu-parser/src/lex.rs)

File: crates/nu-protocol/src/parse_error.rs [hunk 1]
```diff
@@ -413,6 +413,10 @@ pub enum ParseError {
     #[error("{0}")]
     #[diagnostic()]
     LabeledError(String, String, #[label("{1}")] Span),
+
+    #[error("{0}")]
+    #[diagnostic(help("{1}"))]
+    LabeledErrorWithHelp(String, String, String, #[label("{2}")] Span),
```
Comment (sholderbach) [inline crates/nu-protocol/src/parse_error.rs:419]: Can we make this a named struct variant instead of a tuple struct variant, else it is hard to follow what goes where.
```
    LabeledErrorWithHelp{
    	msg: String, 
    	help: String, 
    	label: String, 
    	#[label("{label}")] 
    	span: Span
    	},
```

Comment (1Kinoti) [reply crates/nu-protocol/src/parse_error.rs]: fixed

Comment (sholderbach) [reply crates/nu-protocol/src/parse_error.rs]: Awesome thanks for your quick turnaround!

**Comment by fdncred** [comment]
@1Kinoti we can probably land this if you fix the conflicts please.

**Comment by 1Kinoti** [comment]
@pingiun, i cannot tell why [this](https://github.com/nushell/nushell/blob/2a484a3e7e169e699766e1883ccb1286b2383212/crates/nu-command/tests/commands/def.rs#LL146C1-L151C1) test keeps on failing, i see it was introduced in #8956 but before that, this pr passed all CI. can you maybe tell me what i am doing wrong?

**Comment by pingiun** [comment]
> @pingiun, i cannot tell why [this](https://github.com/nushell/nushell/blob/2a484a3e7e169e699766e1883ccb1286b2383212/crates/nu-command/tests/commands/def.rs#LL146C1-L151C1) test keeps on failing, i see it was introduced in #8956 but before that, this pr passed all CI. can you maybe tell me what i am doing wrong?

Just on the face of it I think you reverted some of my changes. I got this error before I changed the parser correctly. That is the rough idea but I'll check your changes to see where it goes wrong

File: crates/nu-parser/src/parse_keywords.rs [hunk 2]
```diff
@@ -170,7 +180,7 @@ pub fn parse_def_predecl(working_set: &mut StateWorkingSet, spans: &[Span]) {
                 working_set.error(ParseError::DuplicateCommandDef(spans[1]));
             }
         }
-    } else if name == b"extern" && spans.len() >= 3 {
+    } else if decl_name == b"extern" && spans.len() == 3 {
```
Comment (pingiun) [inline crates/nu-parser/src/parse_keywords.rs:183]: This should be >=

**Review by pingiun — CHANGES_REQUESTED** [review]
I believe this is where your error comes from. Externs can now have more than 3 spans

File: crates/nu-parser/src/parse_keywords.rs (hunk 2 above)
Comment (pingiun) [inline crates/nu-parser/src/parse_keywords.rs:183]:
```suggestion
    } else if decl_name == b"extern" && spans.len() >= 3 {
```

Comment (fdncred) [reply crates/nu-parser/src/parse_keywords.rs]: I committed that change. let's see. thanks @pingiun !!

**Comment by 1Kinoti** [comment]
aah, because `extern`s can now have bodies! thanks @pingiun

**Comment by fdncred** [comment]
let's move forward with this. thanks.
